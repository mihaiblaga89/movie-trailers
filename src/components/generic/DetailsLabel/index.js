import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'semantic-ui-react';

const DetailsLabel = ({ label, children }) => {
    return (
        <List.Item>
            <List.Header>{label}</List.Header>
            {Array.isArray(children) ? children.join(', ') : children}
        </List.Item>
    );
};

DetailsLabel.propTypes = {
    label: PropTypes.string.isRequired,
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
};

export default DetailsLabel;
