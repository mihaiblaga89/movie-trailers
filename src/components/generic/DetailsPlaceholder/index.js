import React from 'react';
import { Placeholder, Grid } from 'semantic-ui-react';

const DetailsPlaceholder = () => {
    return (
        <Grid stackable>
            <Grid.Row columns={2}>
                <Grid.Column>
                    <Placeholder fluid>
                        <Placeholder.Header>
                            <Placeholder.Line />
                        </Placeholder.Header>
                        <Placeholder.Paragraph>
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                        </Placeholder.Paragraph>
                        <Placeholder.Paragraph>
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                        </Placeholder.Paragraph>
                        <Placeholder.Paragraph>
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                        </Placeholder.Paragraph>
                        <Placeholder.Paragraph>
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                            <Placeholder.Line />
                        </Placeholder.Paragraph>
                    </Placeholder>
                </Grid.Column>
                <Grid.Column>
                    <Placeholder fluid>
                        <Placeholder.Image rectangular />
                    </Placeholder>
                </Grid.Column>
            </Grid.Row>
        </Grid>
    );
};

export default DetailsPlaceholder;
