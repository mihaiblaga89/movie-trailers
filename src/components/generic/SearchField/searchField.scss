.search-field {
    div.input {
        width: 100%;
    }
}
