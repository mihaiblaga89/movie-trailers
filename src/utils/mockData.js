export default {
    detailsList: {
        genres: [],
        production_companies: [],
        budget: 25,
        production_countries: [],
        release_date: '',
        runtime: 32,
        status: '',
        vote_average: 5,
        vote_count: 232,
        first_air_date: '',
        created_by: [],
        episode_run_time: [],
        networks: [],
        seasons: [],
        last_episode_to_air: null,
        next_episode_to_air: null,
        number_of_episodes: 24,
    },
};
