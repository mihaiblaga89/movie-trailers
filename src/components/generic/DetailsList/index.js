/* eslint-disable camelcase */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { List, Rating } from 'semantic-ui-react';

import DetailsLabel from '../DetailsLabel';

const DetailsList = ({ data, isMovie }) => {
    const {
        genres,
        production_companies,
        budget,
        production_countries,
        release_date,
        runtime,
        status,
        vote_average,
        vote_count,
        first_air_date,
        created_by,
        episode_run_time,
        networks,
        seasons,
        last_episode_to_air,
        next_episode_to_air,
        number_of_episodes,
    } = data;

    // generate an episode string like S01E08 based on the episode object
    const composeEpisodeInfo = episode => {
        const { air_date, episode_number, name, season_number } = episode;
        const pad = toPad => toPad.toString().padStart(2, '0');
        return `${name} - S${pad(season_number)}E${pad(episode_number)} - ${moment(air_date).format('MMMM Do YYYY')}`;
    };

    const renderMovieSpecificData = () => (
        <Fragment>
            {budget ? <DetailsLabel label="Budget">{`$${Number(budget).toLocaleString()}`}</DetailsLabel> : null}
            {production_countries && (
                <DetailsLabel label="Production countries">{production_countries.map(country => country.name)}</DetailsLabel>
            )}
            {release_date && <DetailsLabel label="Release date">{moment(release_date).format('MMMM Do YYYY')}</DetailsLabel>}
            {runtime ? <DetailsLabel label="Runtime">{`${runtime} minutes`}</DetailsLabel> : null}
        </Fragment>
    );

    const renderSeriesSpecificData = () => (
        <Fragment>
            {created_by && <DetailsLabel label="Created by">{created_by.map(creator => creator.name)}</DetailsLabel>}
            {first_air_date && (
                <DetailsLabel label="First air date">{moment(first_air_date).format('MMMM Do YYYY')}</DetailsLabel>
            )}
            {episode_run_time && <DetailsLabel label="Episode run time">{`${episode_run_time[0]} minutes`}</DetailsLabel>}
            {networks && (
                <DetailsLabel label="Networks">
                    {networks.map(network => `${network.name} (${network.origin_country})`)}
                </DetailsLabel>
            )}
            {seasons && (
                <DetailsLabel label="Number of seasons">
                    {seasons.filter(season => season.name.includes('Season')).length.toString()}
                </DetailsLabel>
            )}
            {number_of_episodes ? <DetailsLabel label="Number of episodes">{number_of_episodes.toString()}</DetailsLabel> : null}
            {last_episode_to_air && <DetailsLabel label="Latest episode">{composeEpisodeInfo(last_episode_to_air)}</DetailsLabel>}
            {next_episode_to_air && <DetailsLabel label="Next episode">{composeEpisodeInfo(next_episode_to_air)}</DetailsLabel>}
        </Fragment>
    );

    return (
        <List>
            {vote_average ? (
                <Fragment>
                    <Rating
                        disabled
                        style={{ marginBottom: '25px' }}
                        icon="heart"
                        rating={(vote_average * 5) / 10}
                        maxRating={5}
                    />
                    {vote_count && <span>{`(${vote_count} votes)`}</span>}
                </Fragment>
            ) : null}
            {status && <DetailsLabel label="Status">{status}</DetailsLabel>}
            {genres && <DetailsLabel label="Genres">{genres.map(genre => genre.name)}</DetailsLabel>}
            {production_companies && (
                <DetailsLabel label="Production companies">{production_companies.map(company => company.name)}</DetailsLabel>
            )}
            {isMovie ? renderMovieSpecificData() : renderSeriesSpecificData()}
        </List>
    );
};

DetailsList.propTypes = {
    data: PropTypes.shape({
        genres: PropTypes.array,
        production_companies: PropTypes.array,
        budget: PropTypes.number,
        production_countries: PropTypes.array,
        release_date: PropTypes.string,
        runtime: PropTypes.number,
        status: PropTypes.string,
        vote_average: PropTypes.number,
        vote_count: PropTypes.number,
        first_air_date: PropTypes.string,
        created_by: PropTypes.array,
        episode_run_time: PropTypes.array,
        networks: PropTypes.array,
        seasons: PropTypes.array,
        last_episode_to_air: PropTypes.object,
        next_episode_to_air: PropTypes.object,
        number_of_episodes: PropTypes.number,
    }),
    isMovie: PropTypes.bool.isRequired,
};

export default DetailsList;
