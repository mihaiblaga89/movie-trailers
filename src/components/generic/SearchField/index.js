// eslint keeps wanting me to sort components in a way that will break functionality
/* eslint-disable react/sort-comp */
import React, { Component } from 'react';
import { Search } from 'semantic-ui-react';
import { navigate } from '@reach/router';
import debounce from 'lodash/debounce';

import './searchField.scss';
import API from '../../../services/api';
import { generateSmallPosterURL } from '../../../utils';

class SearchField extends Component {
    state = {
        isLoading: false,
        results: [],
    };

    computeResults = results => {
        return results
            .filter(result => result.media_type !== 'person') // take out persons since we don't support them
            .map(result => ({
                key: result.id,
                title: result.title || result.name,
                image: result.poster_path ? generateSmallPosterURL(result.poster_path) : null,
                description: result.vote_average ? `${result.vote_average} / 10` : null,
                type: result.media_type,
            }));
    };

    reset = () => this.setState({ isLoading: false, results: [] });

    handleSearch = (e, { value }) => {
        this.setState({ isLoading: true });
        if (value.length > 1) {
            API.search(value)
                .then(({ data: { results } }) => this.setState({ results: this.computeResults(results), isLoading: false }))
                .catch(this.reset);
        } else {
            this.reset();
        }
    };

    debouncedSearch = debounce(this.handleSearch, 500, { trailing: true });

    handleResultSelect = (e, { result }) => {
        navigate(`/${result.type === 'tv' ? 'series' : 'movies'}/${result.key}`);
    };

    render() {
        const { isLoading, results } = this.state;
        return (
            <Search
                className="search-field"
                fluid
                loading={isLoading}
                onResultSelect={this.handleResultSelect}
                onSearchChange={this.debouncedSearch}
                results={results}
            />
        );
    }
}

export default SearchField;
