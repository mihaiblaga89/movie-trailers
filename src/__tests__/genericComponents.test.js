import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import DetailsLabel from '../components/generic/DetailsLabel';
import DetailsList from '../components/generic/DetailsList';
import DetailsPlaceholder from '../components/generic/DetailsPlaceholder';
import SearchField from '../components/generic/SearchField';
import mockData from '../utils/mockData';

/* global document */

// keeping the code dry
const tests = [
    { name: 'DetailsLabel', component: <DetailsLabel label="test">Label</DetailsLabel> },
    { name: 'DetailsList', component: <DetailsList isMovie data={mockData.detailsList} /> },
    { name: 'DetailsPlaceholder', component: <DetailsPlaceholder /> },
    { name: 'SearchField', component: <SearchField /> },
];

tests.forEach((test) =>
    it(`${test.name} renders without crashing and matches snapshot`, () => {
        const div = document.createElement('div');
        ReactDOM.render(test.component, div);
        ReactDOM.unmountComponentAtNode(div);
        const tree = renderer.create(test.component).toJSON();
        expect(tree).toMatchSnapshot();
    })
);
