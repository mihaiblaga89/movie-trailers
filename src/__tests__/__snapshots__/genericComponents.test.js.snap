// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsLabel renders without crashing and matches snapshot 1`] = `
<div
  className="item"
  onClick={[Function]}
  role="listitem"
>
  <div
    className="header"
  >
    test
  </div>
  Label
</div>
`;

exports[`DetailsList renders without crashing and matches snapshot 1`] = `
<div
  className="ui list"
  role="list"
>
  <div
    className="ui heart disabled rating"
    onMouseLeave={[Function]}
    role="radiogroup"
    style={
      Object {
        "marginBottom": "25px",
      }
    }
    tabIndex={0}
  >
    <i
      aria-checked={false}
      aria-posinset={1}
      aria-setsize={5}
      className="active icon"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      role="radio"
      tabIndex={-1}
    />
    <i
      aria-checked={false}
      aria-posinset={2}
      aria-setsize={5}
      className="active icon"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      role="radio"
      tabIndex={-1}
    />
    <i
      aria-checked={false}
      aria-posinset={3}
      aria-setsize={5}
      className="icon"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      role="radio"
      tabIndex={-1}
    />
    <i
      aria-checked={false}
      aria-posinset={4}
      aria-setsize={5}
      className="icon"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      role="radio"
      tabIndex={-1}
    />
    <i
      aria-checked={false}
      aria-posinset={5}
      aria-setsize={5}
      className="icon"
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      role="radio"
      tabIndex={-1}
    />
  </div>
  <span>
    (232 votes)
  </span>
  
  <div
    className="item"
    onClick={[Function]}
    role="listitem"
  >
    <div
      className="header"
    >
      Genres
    </div>
    
  </div>
  <div
    className="item"
    onClick={[Function]}
    role="listitem"
  >
    <div
      className="header"
    >
      Production companies
    </div>
    
  </div>
  <div
    className="item"
    onClick={[Function]}
    role="listitem"
  >
    <div
      className="header"
    >
      Budget
    </div>
    $25
  </div>
  <div
    className="item"
    onClick={[Function]}
    role="listitem"
  >
    <div
      className="header"
    >
      Production countries
    </div>
    
  </div>
  
  <div
    className="item"
    onClick={[Function]}
    role="listitem"
  >
    <div
      className="header"
    >
      Runtime
    </div>
    32 minutes
  </div>
</div>
`;

exports[`DetailsPlaceholder renders without crashing and matches snapshot 1`] = `
<div
  className="ui stackable grid"
>
  <div
    className="two column row"
  >
    <div
      className="column"
    >
      <div
        className="ui fluid placeholder"
      >
        <div
          className="header"
        >
          <div
            className="line"
          />
        </div>
        <div
          className="paragraph"
        >
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
        </div>
        <div
          className="paragraph"
        >
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
        </div>
        <div
          className="paragraph"
        >
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
        </div>
        <div
          className="paragraph"
        >
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
        </div>
      </div>
    </div>
    <div
      className="column"
    >
      <div
        className="ui fluid placeholder"
      >
        <div
          className="rectangular image"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchField renders without crashing and matches snapshot 1`] = `
<div
  className="ui fluid search search-field"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="ui icon input"
  >
    <input
      autoComplete="off"
      className="prompt"
      onChange={[Function]}
      onClick={[Function]}
      tabIndex="0"
      type="text"
      value=""
    />
    <i
      aria-hidden="true"
      className="search icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="results transition"
  >
    <div
      className="message empty"
    >
      <div
        className="header"
      >
        No results found.
      </div>
    </div>
  </div>
</div>
`;
